File: Backend/src/handlers/uploadHandler.js
'use strict';

const Fs = require('fs');
const Path = require('path');
const { v4: uuidv4 } = require('uuid');
const config = require('../config');
const logger = require('../utils/logger');

/**
 * Handle file upload
 * @param {Object} file - File object from payload
 * @param {string} type - Type of file (image or video)
 * @returns {Promise<Object>} Upload result
 */
async function handleFileUpload(file, type = 'unknown') {
  return new Promise((resolve, reject) => {
    try {
      // Validate file type
      const fileExtension = Path.extname(file.hapi.filename).toLowerCase();
      const mimeType = file.hapi.headers['content-type'];
      
      let isValidType = false;
      
      if (type === 'image') {
        isValidType = config.uploads.allowedImageTypes.includes(mimeType);
      } else if (type === 'video') {
        isValidType = config.uploads.allowedVideoTypes.includes(mimeType);
      } else {
        // Accept both image and video
        isValidType = [...config.uploads.allowedImageTypes, ...config.uploads.allowedVideoTypes].includes(mimeType);
      }
      
      if (!isValidType) {
        return reject(new Error(`Invalid file type: ${mimeType}. Allowed types for ${type}: ${
          type === 'image' 
            ? config.uploads.allowedImageTypes.join(', ') 
            : config.uploads.allowedVideoTypes.join(', ')
        }`));
      }
      
      // Generate a unique filename
      const filename = `${uuidv4()}${fileExtension}`;
      const filepath = Path.join(config.directories.uploads, filename);
      
      // Create writable stream
      const fileStream = Fs.createWriteStream(filepath);
      
      // Write file
      file.pipe(fileStream);
      
      file.on('end', () => {
        resolve({
          success: true,
          filename,
          filepath,
          originalName: file.hapi.filename,
          mimeType,
          type
        });
      });
      
      file.on('error', (err) => {
        // Clean up file if it was created
        try {
          if (Fs.existsSync(filepath)) {
            Fs.unlinkSync(filepath);
          }
        } catch (e) {
          logger.error('Error cleaning up file after upload failure:', e);
        }
        
        reject(err);
      });
    } catch (err) {
      reject(err);
    }
  });
}

/**
 * Process an uploaded image file
 * @param {string} filepath - Path to the uploaded image file
 * @returns {Promise<Object>} Processing result
 */
async function processImageFile(filepath) {
  // This function would normally process the image for landmark extraction
  // but since we're receiving landmarks directly from the frontend, it's simpler
  const fileStats = Fs.statSync(filepath);
  
  return {
    filepath,
    size: fileStats.size,
    lastModified: fileStats.mtime
  };
}

/**
 * Process an uploaded video file
 * @param {string} filepath - Path to the uploaded video file
 * @returns {Promise<Object>} Processing result
 */
async function processVideoFile(filepath) {
  // This function would normally process the video for landmark extraction
  // but since we're receiving landmarks directly from the frontend, it's simpler
  const fileStats = Fs.statSync(filepath);
  
  return {
    filepath,
    size: fileStats.size,
    lastModified: fileStats.mtime
  };
}

/**
 * Clean up old files in the uploads directory
 * @param {number} maxAgeMinutes - Maximum age of files to keep in minutes
 */
function cleanupUploads(maxAgeMinutes = 60) {
  try {
    const files = Fs.readdirSync(config.directories.uploads);
    const now = Date.now();
    
    files.forEach(file => {
      const filepath = Path.join(config.directories.uploads, file);
      const stats = Fs.statSync(filepath);
      
      // Calculate file age in minutes
      const ageMinutes = (now - stats.mtime.getTime()) / (1000 * 60);
      
      if (ageMinutes > maxAgeMinutes) {
        // Delete old files
        Fs.unlinkSync(filepath);
        logger.info(`Deleted old file: ${filepath}`);
      }
    });
  } catch (err) {
    logger.error('Error cleaning up uploads directory:', err);
  }
}

module.exports = {
  handleFileUpload,
  processImageFile,
  processVideoFile,
  cleanupUploads
};

File: Backend/handlers/upload_handler.py
import os
import aiofiles
from uuid import uuid4
from typing import Dict, Any, Optional, Union
from fastapi import UploadFile, HTTPException, status
from config import Config
import traceback 

def log_info(message: str, meta: Optional[Dict] = None):
    print(f"[INFO] {message}", meta if meta else "")

def log_error(message: str, error: Union[Exception, Dict] = None):
    meta = {"message": str(error), "stack": traceback.format_exc()} if isinstance(error, Exception) else error
    print(f"[ERROR] {message}", meta if meta else "")

class UploadHandler:
    async def handle_file_upload(self, file: UploadFile, file_type: str = 'unknown') -> Dict[str, Any]:
        try:
            file_extension = os.path.splitext(file.filename)[1].lower()
            mime_type = file.content_type

            is_valid_type = False
            if file_type == 'image':
                is_valid_type = mime_type in Config.ALLOWED_IMAGE_TYPES
            elif file_type == 'video':
                is_valid_type = mime_type in Config.ALLOWED_VIDEO_TYPES
            else: # unknown, check both
                is_valid_type = (mime_type in Config.ALLOWED_IMAGE_TYPES or
                                 mime_type in Config.ALLOWED_VIDEO_TYPES)
            
            if not is_valid_type:
                allowed_types_str = ", ".join(Config.ALLOWED_IMAGE_TYPES + Config.ALLOWED_VIDEO_TYPES)
                raise HTTPException(status_code=400, detail=f"Invalid file type: {mime_type}. Allowed types: {allowed_types_str}")
            
            if file.size > Config.MAX_FILE_SIZE_BYTES:
                raise HTTPException(status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE, detail=f"File too large. Max size is {Config.MAX_FILE_SIZE_MB}MB.")

            filename = f"{uuid4()}{file_extension}"
            filepath = os.path.join(Config.UPLOADS_DIR, filename)

            async with aiofiles.open(filepath, "wb") as out_file:
                while content := await file.read(1024 * 1024):
                    await out_file.write(content)

            file_size = os.path.getsize(filepath)

            log_info(f"File uploaded: {filename}", {
                "originalName": file.filename,
                "mimeType": mime_type,
                "type": file_type,
                "size": file_size
            })

            return {
                "success": True,
                "filename": filename,
                "filepath": filepath,
                "originalName": file.filename,
                "mimeType": mime_type,
                "type": file_type
            }
        except HTTPException:
            raise
        except Exception as e:
            log_error("Error uploading file:", e)
            raise HTTPException(status_code=500, detail=f"Error processing file upload: {e}")